include/particle.hpp

```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#if !defined WASH_WSER && !defined WASH_WISB && !defined WASH_WEST && !defined WASH_CSTONE && !defined WASH_WONE
+#error "Please specify an implementation when compiling WASH"
+#endif 
+
+#include <string>
+#include <vector>
+
+#include "vector.hpp"
+
+namespace wash {
+    class Particle {
+    private:
+// All implementations
+        size_t global_idx;
+/// TODO: see if this works if it's a glocal spec defined flag rather than in just one file? 
+#if defined WASH_WONE || defined WASH_CSTONE
```

Comment:
Still not totally sure about conditional compilation like this. I feel like the 'most correct' way of doing this would be inheritance with a different subclass for different WASH_ implemenations. Difficult problem - should be fine doing it this way so far. 

Did we determine whether introduced a clang bug in the end?

Reply:
With regards to subclasses/inheritance - I didn't go down this way initially because there's only ever going to be one version of this in the program so it seemed a bit redundant. 

From some research I did it seemed like what I was doing might've caused a violation of the standard's One Definition Rule. Because when it was included from other files it might not have had the same #defines for whatever reason, meaning clang thought there were two variants one with 2 size_t members and one with one size_t member, used in different files (and hence different translation units) so it didn't throw up an error. The way I fixed this was just to require a compiler level `-DWASH_XXXX` flag so that it would always be defined.